firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,66 +163,101 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
-   *     PersistedFidEntry} is in error state.
-   */
-  private PersistedFidEntry getPersistedFid() throws FirebaseInstallationsException {
-    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
-    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
-      throw new FirebaseInstallationsException(
-          "Failed to get existing fid.", FirebaseInstallationsException.Status.CLIENT_ERROR);
+  private Task<String> addGetIdListener() {
+    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+    StateListener l = new GetIdListener(taskCompletionSource);
+    synchronized (lock) {
+      listeners.add(l);
     }
-    return persistedFidEntry;
+    return taskCompletionSource.getTask();
   }
 
-  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
-    return persistedFidEntry == null
-        || persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTER_ERROR;
+  private Task<InstallationTokenResult> addGetAuthTokenListener(
+      @AuthTokenOption int authTokenOption) {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    StateListener l = new GetAuthTokenListener(utils, taskCompletionSource);
+    synchronized (lock) {
+      if (authTokenOption == FORCE_REFRESH) {
+        shouldRefreshAuthToken = true;
+      }
+      listeners.add(l);
+    }
+    return taskCompletionSource.getTask();
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
-    return t -> {
-      if (t.isSuccessful()) {
-        return (T) t.getResult();
+  private void triggerOnStateReached(PersistedFidEntry persistedFidEntry) {
+    synchronized (lock) {
+      Iterator<StateListener> it = listeners.iterator();
+      while (it.hasNext()) {
+        StateListener l = it.next();
+        boolean doneListening = l.onStateReached(persistedFidEntry);
+        if (doneListening) {
+          it.remove();
+        }
       }
-      return supplier.get();
-    };
+    }
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> awaitFidRegistration(
-      @NonNull Supplier<T> supplier, AwaitListener listener) {
-    return t -> {
-      // Waiting for Task that registers FID on the FIS Servers
-      listener.await(AWAIT_TIMEOUT_IN_SECS, TimeUnit.SECONDS);
-      return supplier.get();
+  private final Runnable doRegistration() {
+    return () -> {
+      try {
+        PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+        // New FID needs to be created
+        if (persistedFidEntry.isErrored() || persistedFidEntry.isNotGenerated()) {
+          String fid = utils.createRandomFid();
+          persistFid(fid);
+          persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+        }
+
+        // Notify the listeners only after force refreshing auth token if shouldRefreshAuthToken is
```

Comment:
trouble is that this means the getId listeners won't be notified until registration happens, which is something we wanted to avoid.

The initial draft version didn't run into this because it called triggerOnStateReached(NOT_REGISTERED) instead of triggerOnStateReached(persistedFidEntry) which meant the listeners were simpler because each step of the state machine always gets called. Could we change it so that it's calling `triggerOnStateReached(newState, persistedFidEntry)` in order to get the step-by-step simplification benefits of the state machine?

Reply:
shouldRefreshAuthToken will be false by default for getId. Hence, getId listener will be notified before registration happens. (Matches the initial draft version).

It is a matter of doing the checks here vs doing the checks in listeners. 

With ```triggerOnStateReached(newState, persistedFidEntry)``` we have to change the definition of both the listeners (because interface changes). ```GetIdListener``` doesn't need newState.